### PR TITLE
Added Error messages on Gateway_Balance and Channel_Authorize

### PR DESCRIPTION
**Pull request by PeterChen13579**
## High Level Overview of Change
Fixed/added error messages for Gateway_balance(issue #4290 | #4548 ) and channel_authorize(#4289 )

### Context of Change
Added error messages when user sends an incorrect request. (Check the above Github issue link for more info.) Everything is under API Version 2 as it is a breaking change.

### Type of Change

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X ] Tests (You added tests for code that already exists, or your new feature included in this PR)

Added own unit test for API Version 2;

### CODE REVIEW

File: src/test/rpc/GatewayBalances_test.cpp [hunk 1]
```diff
@@ -148,6 +148,36 @@ class GatewayBalances_test : public beast::unit_test::suite
...
+    {
+        using namespace std::chrono_literals;
+        using namespace jtx;
+        Env env{
+            *this,
+            envconfig([](std::unique_ptr<Config> c) {
+                c->loadFromString("\n[beta_rpc_api]\n1\n");
+                return c;
+            }),
+            features};
```
Comment (arihantkothari) [inline src/test/rpc/GatewayBalances_test.cpp:162]: Since #4573 is merged, this can probably be changed to Env env{*this}; ?

Comment (PeterChen13579) [reply src/test/rpc/GatewayBalances_test.cpp]: done 👍

File: src/test/app/PayChan_test.cpp [hunk 2]
```diff
@@ -1160,6 +1160,47 @@ struct PayChan_test : public beast::unit_test::suite
...
+        using namespace jtx;
+        using namespace std::literals::chrono_literals;
+
+        Env env{
+            *this,
+            envconfig([](std::unique_ptr<Config> c) {
+                c->loadFromString("\n[beta_rpc_api]\n1\n");
+                return c;
+            }),
+            features};
```
Comment (arihantkothari) [inline src/test/app/PayChan_test.cpp:1176]:
```suggestion
        `Env env{*this, features};`
```

File: src/test/rpc/GatewayBalances_test.cpp (hunk 1 above)
Comment (arihantkothari) [reply src/test/rpc/GatewayBalances_test.cpp]: there's more over here: https://github.com/XRPLF/rippled/pull/4577#discussion_r1239092997

File: src/test/app/PayChan_test.cpp [hunk 3]
```diff
@@ -1160,6 +1160,41 @@ struct PayChan_test : public beast::unit_test::suite
             bob.human());
     }
 
+    void
+    testAccountChannelAuthorizeApiV2(FeatureBitset features)
```
Comment (arihantkothari) [inline src/test/app/PayChan_test.cpp:1164]: Might wanna:

- Add a parameter for api_version.
- Call this method in a loop over versions from apiMinimumSupportedVersion to apiBetaVersion.
- Remove the other method.
- Make the assertion on the error value conditional on the api_version parameter, using condition < 2 for the old behavior and else for the new behavior.

**Review by arihantkothari — CHANGES_REQUESTED** [review]

File: src/test/app/PayChan_test.cpp (hunk 3 above)
Comment (arihantkothari) [reply src/test/app/PayChan_test.cpp:1164]: https://github.com/XRPLF/rippled/pull/4552#discussion_r1242329180

File: src/ripple/rpc/handlers/GatewayBalances.cpp [hunk 4]
```diff
@@ -116,7 +122,18 @@ doGatewayBalances(RPC::JsonContext& context)
...
+            // invalidHotwallet should be used when the account exists, but does
+            // not have currency issued by the account from the request.
+            if (context.apiVersion == 1)
+            {
+                result[jss::error] = "invalidHotWallet";
+            }
+            else
+            {
+                result[jss::error] = "invalidParams";
+            }
```
Comment (arihantkothari) [inline src/ripple/rpc/handlers/GatewayBalances.cpp:136]: Please follow the guideline that was shared before.

```
if (version < FEATURE_X_VERSION) {
  // old behavior
} else {
  // new behavior
}

if (version >= FEATURE_X_VERSION) {
  // new behavior where old behavior was just "do nothing"
}

if (version < FEATURE_X_VERSION) {
  // old behavior where new behavior is "stop doing that"
}
```

Comment (arihantkothari) [reply src/ripple/rpc/handlers/GatewayBalances.cpp:136]: Also consider replacing 1 with 1u. 1 is signed and 1u is unsigned.

**Review by arihantkothari — CHANGES_REQUESTED** [review]

**Comment by intelliot** [comment]
Hi @PeterChen13579 - please [sign your commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification).

**Comment by intelliot** [comment]
Commit signature looks good. Can you confirm that this PR is ready for review?

**Comment by PeterChen13579** [comment]
> Commit signature looks good. Can you confirm that this PR is ready for review?

Yes, it's ready for review 👍

**Review by arihantkothari — APPROVED** [review]

**Comment by intelliot** [comment]
@mounikakun would you be able to review this?

**Comment by mounikakun** [comment]
@intelliot I am not familiar with c++ code, so I cannot review this.

File: src/ripple/rpc/impl/RPCHelpers.h [hunk 5]
```diff
@@ -207,7 +207,10 @@ std::optional<Seed>
 parseRippleLibSeed(Json::Value const& params);
 
 std::pair<PublicKey, SecretKey>
-keypairForSignature(Json::Value const& params, Json::Value& error);
+keypairForSignature(
+    Json::Value const& params,
+    Json::Value& error,
+    std::optional<std::reference_wrapper<JsonContext>> context = std::nullopt);
```
Comment (gregtatcam) [inline src/ripple/rpc/impl/RPCHelpers.h:213]: Consider changing to `std::optional<uint> const& apiVersion`. Actually, maybe we can even do this: `uint apiVersion = RPC::apiVersionIfUnspecified`?

File: src/test/app/PayChan_test.cpp [hunk 6]
```diff
@@ -24,9 +24,9 @@
 #include <ripple/protocol/PayChan.h>
 #include <ripple/protocol/TxFlags.h>
 #include <ripple/protocol/jss.h>
-#include <test/jtx.h>
-
+#include <ripple/rpc/impl/RPCHelpers.h>
```
Comment (gregtatcam) [inline src/test/app/PayChan_test.cpp:27]: Why did you change the headers order? The original order and the new line looks right to me.

File: src/test/app/PayChan_test.cpp [hunk 7]
```diff
@@ -1160,6 +1160,50 @@ struct PayChan_test : public beast::unit_test::suite
...
+        env.close();
+
+        // test for api_version 2
+        Json::Value args{Json::objectValue};
+        args[jss::api_version] = apiVersion;
+        args[jss::channel_id] = chan1Str;
+        args[jss::key_type] = "ed255191";
+        args[jss::seed] = "snHq1rzQoN2qiUkC3XF5RyxBzUtN";
+        args[jss::amount] = 51110000;
+        if (apiVersion < 2u)
```
Comment (gregtatcam) [inline src/test/app/PayChan_test.cpp:1189]: Consider moving the testVersion loop inside this function. The channel above doesn't need to be created twice. Then this is going to be:
>     for (auto apiVersion = RPC::apiMinimumSupportedVersion;
>             apiVersion <= RPC::apiBetaVersion;
>             ++apiVersion)
>        {
>            testcase("PayChan Channel_Auth RPC Api " + std::to_string(apiVersion));
>            args[jss::api_version] = apiVersion;
>            auto const rs = env.rpc(
>                "json",
>                "channel_authorize",
>                args.toStyledString())[jss::result];
>            auto const error = apiVersion < 2u ? "invalidParams" : "badKeyType";
>            BEAST_EXPECT(rs[jss::error] == error);
>        }

File: src/test/app/PayChan_test.cpp [hunk 8]
```diff
@@ -2139,6 +2183,12 @@ struct PayChan_test : public beast::unit_test::suite
         testAccountChannelsRPC(features);
         testAccountChannelsRPCMarkers(features);
         testAccountChannelsRPCSenderOnly(features);
+        for (auto testVersion = RPC::apiMinimumSupportedVersion;
```
Comment (gregtatcam) [inline src/test/app/PayChan_test.cpp:2186]: Consider moving the loop inside the function.

File: src/test/rpc/GatewayBalances_test.cpp [hunk 9]
```diff
@@ -28,123 +29,159 @@ class GatewayBalances_test : public beast::unit_test::suite
 {
 public:
     void
-    testGWB(FeatureBitset features)
+    testGWB(FeatureBitset features, unsigned int apiVersion)
```
Comment (gregtatcam) [inline src/test/rpc/GatewayBalances_test.cpp:32]: This really looks to me like two different tests. Consider keeping the original `testGWB()` as is and adding say `testGWBApiV2(FeatureBitset)`, which has the added code.

File: src/test/rpc/GatewayBalances_test.cpp [hunk 10]
```diff
@@ -207,8 +244,13 @@ class GatewayBalances_test : public beast::unit_test::suite
...
         auto const sa = supported_amendments();
-        testGWB(sa - featureFlowCross);
-        testGWB(sa);
+        for (auto testVersion = RPC::apiMinimumSupportedVersion;
+             testVersion <= RPC::apiBetaVersion;
+             ++testVersion)
+        {
+            testGWB(sa - featureFlowCross, testVersion);
+            testGWB(sa, testVersion);
+        }
```
Comment (gregtatcam) [inline src/test/rpc/GatewayBalances_test.cpp:253]: Per above suggestion this then can be changed to:
>       for (auto feature: {sa - featureFlowCross, sa})
>       {
>           testGWB(feature);
>           testGWBApiV2(feature);
>       }

**Review by gregtatcam — APPROVED** [review]
👍 LGTM

**Comment by PeterChen13579** [comment]
@gregtatcam Thank you so much for reviewing my PR. I really appreciate it 😄

**Comment by gregtatcam** [comment]
> @gregtatcam Thank you so much for reviewing my PR. I really appreciate it 😄

You are welcome!

File: src/ripple/protocol/impl/ErrorCodes.cpp [hunk 11]
```diff
@@ -76,6 +76,7 @@ constexpr static ErrorInfo unorderedErrorInfos[]{
     {rpcINTERNAL,               "internal",             "Internal error.", 500},
     {rpcINVALID_LGR_RANGE,      "invalidLgrRange",      "Ledger range is invalid.", 400},
     {rpcINVALID_PARAMS,         "invalidParams",        "Invalid parameters.", 400},
+    {rpcINVALID_HOTWALLET,      "invalidHotWallet",     "Invalid hotwalloet.", 400},
```
Comment (shawnxie999) [inline src/ripple/protocol/impl/ErrorCodes.cpp:79]: `Invalid hotwalloet.`

is this a typo?

File: src/ripple/rpc/impl/RPCHelpers.cpp [hunk 12]
```diff
@@ -900,7 +903,14 @@ keypairForSignature(Json::Value const& params, Json::Value& error)
...
         {
-            error = RPC::invalid_field_error(jss::key_type);
+            if (apiVersion > 1u)
+            {
+                error = RPC::make_error(rpcBAD_KEY_TYPE);
+            }
+            else
+            {
+                error = RPC::invalid_field_error(jss::key_type);
+            }
```
Comment (shawnxie999) [inline src/ripple/rpc/impl/RPCHelpers.cpp:913]: are the brackets needed here?

File: src/ripple/rpc/handlers/GatewayBalances.cpp [hunk 13]
```diff
@@ -116,7 +122,18 @@ doGatewayBalances(RPC::JsonContext& context)
...
+            // invalidHotwallet should be used when the account exists, but does
+            // not have currency issued by the account from the request.
+            if (context.apiVersion < 2u)
+            {
+                RPC::inject_error(rpcINVALID_HOTWALLET, result);
+            }
+            else
+            {
+                RPC::inject_error(rpcINVALID_PARAMS, result);
+            }
```
Comment (shawnxie999) [inline src/ripple/rpc/handlers/GatewayBalances.cpp:136]: brackets here as well

**Comment by intelliot** [comment]
- Replaced by https://github.com/XRPLF/rippled/pull/4618
